contrib_bots/lib/gcalendar.py

```diff
@@ -0,0 +1,367 @@
+
+    Returns:
+        Values, a tuple that contains the 2 (or 3) parameters of the message
+    '''
+    content = message.replace('@gcalendar ', '')
+
+    try:
+        splits = content.split('|')
+        title = splits[0].strip()
+
+        settings = {
+            'RETURN_AS_TIMEZONE_AWARE': True
+        }
+
+        if len(splits) == 4: # Delimited event with timezone
+            settings['TIMEZONE'] = splits[3].strip()
+
+        start_date = dateparser.parse(splits[1].strip(), settings=settings)
+        end_date = None
+
+        if len(splits) >= 3: # Delimited event
+            end_date = dateparser.parse(splits[2].strip(), settings=settings)
+
+        if not start_date or (len(splits) >= 3 and not end_date):  # Unknown date format
+            raise MessageParseError(2)
+
+        # Notice there isn't a "full day event with timezone", because the
+        # timezone is irrelevant in that type of events
+    except IndexError as e:
+        logging.exception(e)
+        return MessageParseError(1), None, None
+    except MessageParseError as e:
+        logging.exception(e)
+        return e, None, None
+    except UnknownTimeZoneError as e:
+        logging.exception(e)
+        return MessageParseError(3), None, None
+
+    return title, start_date, end_date
+
+class GCalendarHandler(object):
+    '''
+    This plugin facilitates creating Google Calendar events.
+
+    Usage:
+        For delimited events:
+            @gcalendar <event_title> | <start_date> | <end_date> | <timezone> (optional)
+        For full-day events:
+            @gcalendar <event_title> | <start_date>
+
+    The "event-title" supports all characters but pipes (|)
+
+    Timezones can be specified in both numerical (+00:00) or abbreviated
+    format (UTC).
+    If no timezone is specified, the server's will be considered as default.
+
+    Right now it only works for the calendar set as "primary" in your account.
+    '''
+    def __init__(self):
+        credentials = get_credentials()
+        http = credentials.authorize(httplib2.Http())
+        self.service = discovery.build('calendar', 'v3', http=http)
+
+    def usage(self):
+        return '''
+            This plugin will allow users to create events
+            in a Google Calendar. Users should preface
+            messages with "@gcalendar".
+
+            Before running this, make sure to register a new
+            project in the Google Developers Console
+            (https://console.developers.google.com/start/api?id=calendar)
+            and download "client secret" as a JSON file,
+            and save it to ~/.gcal_client_secret.json.
+            '''
+
+    def triage_message(self, message, client):
+        original_content = message['content']
+
+        is_gcalendar = original_content.startswith('@gcalendar')
+
+        return is_gcalendar
+
+    def handle_message(self, message, client, state_handler):
```

Comment:
But instead of separate functions I'd go with separate methods. Because all of them are bots-related.